### PR TITLE
add gd32e103xx pin config

### PR DESCRIPTION
**Pull request by feilongfl**
1. add pinconfig yaml for gd32e103xx
2. gen files by `gd32pinctrl.py` script.

I noticed that the serial name of Gigadevice has changed unusually a few months ago.
They removed the CAN-FD module of the gd32e103 chip, 
and the old `gd32e103` with CAN-FD was renamed as `gd32c103`.
This PR for the new `gd32e103` only.

### CODE REVIEW

File: pinconfigs/gd32e103xx.yml [hunk 1]
```diff
@@ -0,0 +1,789 @@
...
+    memories: [B, 8]
+
+signal-configs:
+  ADC01_IN0:
+    modes: [analog]
+  ADC01_IN1:
+    modes: [analog]
+  ADC01_IN10:
+    modes: [analog]
+    exclude-pincodes: ['T', 'C']
```
Comment (gmarull) [inline pinconfigs/gd32e103xx.yml:42]: No need to use `'T'`, just use `T`

File: pinconfigs/gd32e103xx.yml [hunk 2]
```diff
@@ -0,0 +1,789 @@
...
+# - 64 pins:  R
+# - 48 pins:  C
+# - 36 pins:  T
+#
+# Memory codes:
+#
+# - 128Kb Flash, 32Kb SRAM: B
+# - 64Kb  Flash, 20Kb SRAM: 8
+#
+# SPDX-License-Identifier: Apache 2.0
```
Comment (gmarull) [inline pinconfigs/gd32e103xx.yml:19]: missing copyright

**Review by gmarull — CHANGES_REQUESTED** [review]

File: include/dt-bindings/pinctrl/gd32e103t(b-8)xx-pinctrl.h [hunk 3]
```diff
@@ -0,0 +1,423 @@
+/*
+ * Autogenerated file
+ *
+ * SPDX-License-Identifier: Apache 2.0
+ */
+
+#include "gd32e103xx-afio.h"
```
Comment (gmarull) [inline include/dt-bindings/pinctrl/gd32e103t(b-8)xx-pinctrl.h:7]: this file is missing in this PR

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @feilongfl ,
Thank you for your PR. Please, link this with your board at Zephyr main repo.
We need build a board there using this newer changes to accept PR.

**Comment by feilongfl** [comment]
> Hi @feilongfl , Thank you for your PR. Please, link this with your board at Zephyr main repo. We need build a board there using this newer changes to accept PR.

Sorry for waste your guys time to review this in-process PR.
When I use this one to compile soc, I found I miss something must be done here. I'll push this PR when I fix error.

**Comment by nandojve** [comment]
> > Hi @feilongfl , Thank you for your PR. Please, link this with your board at Zephyr main repo. We need build a board there using this newer changes to accept PR.
> 
> Sorry for waste your guys time to review this in-process PR. When I use this one to compile soc, I found I miss something must be done here. I'll push this PR when I fix error.

Don't worry about it @feilongfl everyone is learning here.
Please, update your PRs at Zephyr main repo with missing info to be possible move forward.

You can look https://github.com/zephyrproject-rtos/zephyr/pull/40283 and use it as guidance.

Tips:

- No need to add wip, when you open PR, simple select Draft when you create it. This tell us that PR is still in progress. If you need assistance, you can ask and someone will, eventually, help you.
- After people let you reviews, you can ask to them look again by clicking at circle with arrows, left side people names. That will inform reviewers that you already update everything or need that person check again PR.

**Comment by feilongfl** [comment]
Hi @nandojve  @gmarull 
I've update this PR, and tested `sample/hello_world` with https://github.com/zephyrproject-rtos/zephyr/pull/36833

- resolved problem which raised by @gmarull 
- add `gd32e10x_libopt.h` from `GD32E10x_Firmware_Library_V1.2.0/Examples/GPIO/Running_led/gd32e10x_libopt.h`
- Some patches in HAL
  - force use external 8M oscillator
      I think we will use value from devicetree later, so there only fix for `gd32e103v_eval` board.
  - remove nvic config
     I think nvic is config by zephyr, don't need to config here.
  - enable MPU flag.
    I found that all of header file in `hal_gigadevice` module is include only when MPU is enable.
    When the MPU is not available, I don't know how to include the files.
    so I think temporary enable `__MPU_PRESENT` marco is a good idea.

File: pinconfigs/gd32e103xx.yml [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+    modes: [out]
+  USBFS_VBUS:
+    modes: [inp]
+  WKUP:
+    modes: [inp]
+
+pins:
+  PA0:
+    pincodes: [C, V, R, T]
+    afs: ['TIMER7_ETI', 'ADC01_IN0', 'TIMER1_CH0_ETI', 'TIMER4_CH0', 'USART1_CTS', 'WKUP']
```
Comment (gmarull) [inline pinconfigs/gd32e103xx.yml:453]: no need for `''` quoting, see other files

File: pinconfigs/gd32e103xx.yml [hunk 5]
```diff
@@ -0,0 +1,790 @@
...
+    afs: ['TRACED3', 'EXMC_A22'] 
+  PE7:
+    pincodes: [V]
+    afs: ['EXMC_D4'] 
+  PE8:
+    pincodes: [V]
+    afs: ['EXMC_D5'] 
+  PE9:
+    pincodes: [V]
+    afs: ['EXMC_D6']
```
Comment (gmarull) [inline pinconfigs/gd32e103xx.yml:690]: looks like you're missing pins: what about PE10..PE15?

File: include/dt-bindings/pinctrl/gd32e103c(b-8)xx-pinctrl.h [hunk 6]
```diff
@@ -0,0 +1,689 @@
+/*
+ * Autogenerated file
+ *
+ * SPDX-License-Identifier: Apache 2.0
+ */
+
+#include "gd32e103xx-afio.h"
```
Comment (gmarull) [inline include/dt-bindings/pinctrl/gd32e103c(b-8)xx-pinctrl.h:7]: where is this file?

File: include/dt-bindings/pinctrl/gd32e103xx-afio.h [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+/** TIMER12 (no remap) */
+#define GD32_TIMER12_NORMP	GD32_REMAP(1U, 8U, 0x1U, 0U)
+/** TIMER12 (remap) */
+#define GD32_TIMER12_RMP	GD32_REMAP(1U, 8U, 0x1U, 1U)
+
+/** TIMER13 (no remap) */
+#define GD32_TIMER13_NORMP	GD32_REMAP(1U, 9U, 0x1U, 0U)
+/** TIMER13 (remap) */
+#define GD32_TIMER13_RMP	GD32_REMAP(1U, 9U, 0x1U, 1U)
```
Comment (gmarull) [inline include/dt-bindings/pinctrl/gd32e103xx-afio.h:97]: these are not available on E10X, please, check datasheet carefully.

File: gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h [hunk 8]
```diff
@@ -49,17 +49,8 @@ OF SUCH DAMAGE.
...
-  #define HXTAL_VALUE_8M  HXTAL_VALUE
-  #elif defined(GD32E103R_START)
-  #define HXTAL_VALUE    ((uint32_t)25000000) /*!< value of the external oscillator in Hz */
-  #define HXTAL_VALUE_25M  HXTAL_VALUE
-  #else
-  #error "Please select the target board type used in your application (in gd32e10x.h file)"
-  #endif
-#endif /* high speed crystal oscillator value */
+#define HXTAL_VALUE_8M    ((uint32_t)8000000)
+#define HXTAL_VALUE       HXTAL_VALUE_8M /*!< value of the external oscillator in Hz */
```
Comment (gmarull) [inline gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h:53]: why is this change required? please, check https://github.com/zephyrproject-rtos/zephyr/pull/34970/commits/a93908563631ddba6308587e77cffd70bde30411 on how to solve this properly

File: gd32e10x/cmsis/gd/gd32e10x/source/system_gd32e10x.c [hunk 9]
```diff
@@ -130,11 +130,11 @@ void SystemInit (void)
...
-#ifdef VECT_TAB_SRAM
-  nvic_vector_table_set(NVIC_VECTTAB_RAM,VECT_TAB_OFFSET);
-#else
-  nvic_vector_table_set(NVIC_VECTTAB_FLASH,VECT_TAB_OFFSET);
-#endif
+// #ifdef VECT_TAB_SRAM
+//   nvic_vector_table_set(NVIC_VECTTAB_RAM,VECT_TAB_OFFSET);
+// #else
+//   nvic_vector_table_set(NVIC_VECTTAB_FLASH,VECT_TAB_OFFSET);
+// #endif
```
Comment (gmarull) [inline gd32e10x/cmsis/gd/gd32e10x/source/system_gd32e10x.c:137]: please, only C style comments `/* */`

File: gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h [hunk 10]
```diff
@@ -94,7 +94,8 @@ OF SUCH DAMAGE.
 
 /* configuration of the Cortex-M4 processor and core peripherals */
 #define __CM4_REV                 0x0001   /*!< Core revision r0p1                                       */
-#define __MPU_PRESENT             0        /*!< GD32E10x do not provide MPU                              */
+// todo: change __MPU_PRESENT to 0
+#define __MPU_PRESENT             1        /*!< GD32E10x do not provide MPU                              */
```
Comment (gmarull) [inline gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h:98]: why is this required?

**Review by gmarull — CHANGES_REQUESTED** [review]

File: pinconfigs/gd32e103xx.yml (hunk 5 above)
Comment (feilongfl) [reply pinconfigs/gd32e103xx.yml:690]: PE10..PE15 at line 649-666.
I'll change the pin order.

File: include/dt-bindings/pinctrl/gd32e103xx-afio.h [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+/** TIMER3 (remap) */
+#define GD32_TIMER3_RMP		GD32_REMAP(0U, 12U, 0x1U, 1U)
+
+/** CAN function is not available in this chip due to hardware issues */
+/** CAN0 (no remap) */
+#define GD32_CAN0_NORMP		GD32_REMAP(0U, 13U, 0x3U, 0U)
+/** CAN0 (partial remap) */
+#define GD32_CAN0_PRMP		GD32_REMAP(0U, 13U, 0x3U, 2U)
+/** CAN0 (full remap) */
+#define GD32_CAN0_FRMP		GD32_REMAP(0U, 13U, 0x3U, 3U)
```
Comment (nandojve) [inline include/dt-bindings/pinctrl/gd32e103xx-afio.h:72]: I think if it is not available it shouldn't be defined. Please drop CAN related.

File: include/dt-bindings/pinctrl/gd32e103xx-afio.h [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+/** TIMER4CH3 (no remap) */
+#define GD32_TIMER4CH3_NORMP   GD32_REMAP(0U, 16U, 0x1U, 0U)
+/** TIMER4CH3 (remap) */
+#define GD32_TIMER4CH3_RMP     GD32_REMAP(0U, 16U, 0x1U, 1U)
+
+/** CAN function is not available in this chip due to hardware issues */
+/** CAN1 (no remap) */
+#define GD32_CAN1_NORMP		GD32_REMAP(0U, 22U, 0x1U, 0U)
+/** CAN1 (remap) */
+#define GD32_CAN1_RMP		GD32_REMAP(0U, 22U, 0x1U, 1U)
```
Comment (nandojve) [inline include/dt-bindings/pinctrl/gd32e103xx-afio.h:83]: Same...

**Review by nandojve — APPROVED** [review]
Hi @feilongfl ,

Nice work!

I think you only need to drop all CAN related at gd32e103.

File: gd32e10x/cmsis/gd/gd32e10x/source/system_gd32e10x.c [hunk 13]
```diff
@@ -129,12 +129,14 @@ void SystemInit (void)
 
     /* configure the system clock source, PLL Multiplier, AHB/APBx prescalers and Flash settings */
     system_clock_config();
-    
+
+/*
```
Comment (gmarull) [inline gd32e10x/cmsis/gd/gd32e10x/source/system_gd32e10x.c:133]: nit: trailing spaces here

File: gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h [hunk 14]
```diff
@@ -50,10 +50,10 @@ OF SUCH DAMAGE.
 
 /* define value of high speed crystal oscillator (HXTAL) in Hz */
 #if !defined  HXTAL_VALUE
-  #ifdef GD32E103V_EVAL
+  #ifdef CONFIG_BOARD_GD32E103V_EVAL
   #define HXTAL_VALUE    ((uint32_t)8000000) /*!< value of the external oscillator in Hz */
   #define HXTAL_VALUE_8M  HXTAL_VALUE
-  #elif defined(GD32E103R_START)
+  #elif defined(CONFIG_BOARD_GD32E103R_START)
```
Comment (gmarull) [inline gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h:56]: HAL headers can't depend on Kconfig variables. Instead, Zephyr build system should insert the `HXTAL_VALUE` based on a Kconfig or DT value. See https://github.com/zephyrproject-rtos/zephyr/commit/a93908563631ddba6308587e77cffd70bde30411

**Review by gmarull — CHANGES_REQUESTED** [review]

File: gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h (hunk 14 above)
Comment (feilongfl) [reply gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h]: I tried to use HXTAL_VALUE, but I'm a bit unsure if it's the same as your idea.

File: gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h (hunk 10 above)
Comment (feilongfl) [reply gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h:115]: I use this command to check header include
```
/opt/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -H -E -DBUILD_VERSION=v2.7.99-2019-gbe7336a1ef94 -DGD32E10X -DKERNEL -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/feilong/Program-ext/zephyr/zephyr/include -I/home/feilong/Program-ext/zephyr/zephyr/build/zephyr/include/generated -I/home/feilong/Program-ext/zephyr/zephyr/soc/arm/gigadevice/gd32e10x -I/home/feilong/Program-ext/zephyr/zephyr/soc/arm/gigadevice/common/. -I/home/feilong/Program-ext/zephyr/zephyr/soc/arm/gigadevice/gd32e10x/. -I/home/feilong/Program-ext/zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/home/feilong/Program-ext/zephyr/modules/hal/gigadevice/gd32e10x/cmsis/gd/gd32e10x/include -I/home/feilong/Program-ext/zephyr/modules/hal/gigadevice/gd32e10x/standard_peripheral/include -I/home/feilong/Program-ext/zephyr/modules/hal/gigadevice/include -isystem /home/feilong/Program-ext/zephyr/zephyr/lib/libc/minimal/include -isystem /opt/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include -isystem /opt/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include-fixed -Os -imacros /home/feilong/Program-ext/zephyr/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee -imacros /home/feilong/Program-ext/zephyr/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/feilong/Program-ext/zephyr/zephyr/samples/basic/minimal=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/feilong/Program-ext/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/feilong/Program-ext/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -o build/e103.i -c /home/feilong/Program-ext/zephyr/zephyr/drivers/pinctrl/pinctrl_gd32_afio.c 2>build/header.txt
```

then I find the include chain of hal_gigadevice is:
```
. /home/feilong/Program-ext/zephyr/zephyr/include/drivers/pinctrl.h
.. /home/feilong/Program-ext/zephyr/zephyr/include/device.h
... /home/feilong/Program-ext/zephyr/zephyr/include/init.h
.... /home/feilong/Program-ext/zephyr/zephyr/include/kernel.h
..... /home/feilong/Program-ext/zephyr/zephyr/include/kernel_includes.h
...... /home/feilong/Program-ext/zephyr/zephyr/include/arch/cpu.h
....... /home/feilong/Program-ext/zephyr/zephyr/include/arch/arm/aarch32/arch.h
........ /home/feilong/Program-ext/zephyr/zephyr/include/arch/arm/aarch32/mpu/arm_mpu.h
......... /home/feilong/Program-ext/zephyr/zephyr/include/arch/arm/aarch32/mpu/arm_mpu_v7m.h
.......... /home/feilong/Program-ext/zephyr/zephyr/include/arch/arm/aarch32/cortex_m/cmsis.h
........... /home/feilong/Program-ext/zephyr/zephyr/soc/arm/gigadevice/gd32e10x/soc.h
............ /home/feilong/Program-ext/zephyr/modules/hal/gigadevice/gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h
............. /home/feilong/Program-ext/zephyr/modules/hal/gigadevice/gd32e10x/standard_peripheral/include/gd32e10x_libopt.h
.............. /home/feilong/Program-ext/zephyr/modules/hal/gigadevice/gd32e10x/standard_peripheral/include/gd32e10x_gpio.h
.............. /home/feilong/Program-ext/zephyr/modules/hal/gigadevice/gd32e10x/standard_peripheral/include/gd32e10x_usart.h

```

because of the code in https://github.com/zephyrproject-rtos/zephyr/blob/aceed99e2d224960d398fd812032ba71eba10162/include/arch/arm/aarch32/arch.h#L187-L197
if I don't enable MPU, It will not include any header file in `hal_gigadevice`.

**Review by nandojve — CHANGES_REQUESTED** [review]
Drop `hal:gd32e10x: temporary enable MPU flag` commit. You need remove MPU at dts and soc definition.
Drop `standard_peripheral/include/gd32e10x_fmc.c` file.

**Comment by feilongfl** [comment]
> Drop `hal:gd32e10x: temporary enable MPU flag` commit. You need remove MPU at dts and soc definition.

If I Drop this commit, I'll get a compile error that header file in hal_gigadevice is not included.

I try to analyse why these file not included, I write them here.
https://github.com/zephyrproject-rtos/hal_gigadevice/pull/12#discussion_r767031750

Currently I have found no way to solve this problem other than turning on MPU.

**Comment by nandojve** [comment]
> 

Please, check https://github.com/zephyrproject-rtos/zephyr/pull/36833#pullrequestreview-829441476 and let me know result.

**Comment by feilongfl** [comment]
> > 
> 
> Please, check [zephyrproject-rtos/zephyr#36833 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/36833#pullrequestreview-829441476) and let me know result.

You're right, I forget that I can also modify the driver code.

File: gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h (hunk 10 above)
Comment (feilongfl) [reply gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h]: fixed by https://github.com/zephyrproject-rtos/hal_gigadevice/pull/12#issuecomment-991413500

File: gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h [hunk 15]
```diff
@@ -57,7 +57,15 @@ OF SUCH DAMAGE.
...
-  #error "Please select the target board type used in your application (in gd32e10x.h file)"
+    /* get external oscillator from KCONFIG */
+    #define HXTAL_VALUE CONFIG_GD32_HXTAL_VALUE
+    #if HXTAL_VALUE == 8000000
+      #define HXTAL_VALUE_8M  HXTAL_VALUE
+    #elif HXTAL_VALUE == 25000000
+      #define HXTAL_VALUE_25M  HXTAL_VALUE
+    #else
+      #error "GD32E10X lib only support 8M and 25M oscillator (HXTAL)"
+    #endif
```
Comment (gmarull) [inline gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h:68]: There's no need to do that, just drop this commit and do the same as https://github.com/zephyrproject-rtos/zephyr/commit/a93908563631ddba6308587e77cffd70bde30411 (cherry-pick it in your PR while it's not merged, even though it's now approved so it'll go in soon)

File: gd32e10x/standard_peripheral/include/gd32e10x_fmc.c [hunk 16]
```diff
@@ -1,880 +0,0 @@
-/*!
```
Comment (gmarull) [inline gd32e10x/standard_peripheral/include/gd32e10x_fmc.c:1]: why this commit? just fixup HAL addition commit

**Review by gmarull — CHANGES_REQUESTED** [review]

File: gd32e10x/standard_peripheral/include/gd32e10x_fmc.c (hunk 16 above)
Comment (nandojve) [reply gd32e10x/standard_peripheral/include/gd32e10x_fmc.c:1]: gd32e10x SoC is already at hal_gigadevice and by unknown reason gd32e10x_fmc.c is duplicated at include dir.
So it is safe drop this file.

**Review by nandojve — CHANGES_REQUESTED** [review]

File: gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h (hunk 15 above)
Comment (feilongfl) [reply gd32e10x/cmsis/gd/gd32e10x/include/gd32e10x.h]: Oh,I see `HXTAL_VALUE` was defined in cmake script, I'll remove it.
Because of code in https://github.com/zephyrproject-rtos/hal_gigadevice/blob/main/gd32e10x/cmsis/gd/gd32e10x/source/system_gd32e10x.c#L745-L751 ,
I need `HXTAL_VALUE_8M` or `HXTAL_VALUE_25M` marco to make clock config success.

**Review by gmarull — APPROVED** [review]
